<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [731]:
import pandas as pd
import psycopg2

In [733]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [734]:
# текст запроса
query_3_1 = f'''
SELECT 
    COUNT(id)
FROM public.VACANCIES
'''

In [735]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\1114379196.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [736]:
# текст запроса
query_3_2 = f'''
SELECT 
    COUNT(id) 
FROM public.EMPLOYERS
'''

In [737]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\3062646138.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [738]:
# текст запроса
query_3_3 = f'''
SELECT 
    COUNT(id) 
FROM public.AREAS
'''

In [739]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\928149064.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [740]:
# текст запроса
query_3_4 = f'''
SELECT 
    COUNT(id) 
FROM public.INDUSTRIES
'''

In [741]:
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\2172479966.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    294

***

# Выводы по предварительному анализу данных
Предварительный анализ данных показал, что в базе содержатся:
* 49197 вакансий;
* 23501 работодателей;
* 1362 городов;
* 294 сфер деятельностей. 


# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [742]:
# текст запроса
query_4_1 = f'''
SELECT 
    COUNT(v.id) cnt,
    a.name area
FROM public.VACANCIES v
    JOIN public.AREAS a ON v.area_id = a.id
GROUP BY 2
ORDER BY 1 DESC
'''

In [743]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\2457173557.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt             area
0    5333           Москва
1    2851  Санкт-Петербург
2    2112            Минск
3    2006      Новосибирск
4    1892           Алматы
..    ...              ...
764     1       Тарко-Сале
765     1    Новоаннинский
766     1            Бирск
767     1           Сасово
768     1             Анна

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [744]:
# текст запроса
query_4_2 = f'''
SELECT 
    COUNT(*)
FROM public.VACANCIES
WHERE salary_from IS NOT NULL 
    OR salary_to IS NOT NULL
'''

In [745]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\970655580.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [746]:
# текст запроса
query_4_3 = f'''
SELECT 
    ROUND(AVG(salary_from)) avg_salary_from,
    ROUND(AVG(salary_to)) avg_salary_to
FROM public.VACANCIES
'''

In [747]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\2500607024.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [748]:
# текст запроса
query_4_4 = f'''
SELECT 
    COUNT(*),
    schedule,
    employment
FROM public.VACANCIES
GROUP BY schedule, employment
ORDER BY COUNT(*) DESC
'''

In [749]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\2022935265.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [750]:
# текст запроса
query_4_5 = f'''
SELECT 
    COUNT(id) cnt,
    experience
FROM public.VACANCIES
WHERE experience IS NOT NULL
GROUP BY experience
ORDER BY COUNT(id) 
'''

In [751]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\2316260449.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt          experience
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет

***

# Выводы по детальному анализу вакансий
* Больше всего вакансий в таких городах, как Москва, Санкт-Петербург и Новосибирск. 8 строчек из топ-10 (см. Дополнительный запрос №1) занимают российские города-миллионники. Из этого можно сделать вывод, что чем больше население города, тем больше количество предлагаемых вакансий. Также в топ-5 попали 1 город (Минск) из Беларуси и 1 город (Алматы) из Казахстана, что говорит о том, что работодатели из других стран также пользуются сервисом.
* Общее количество вакансий - 49197, примерно в половине (24073) вакансий не указана либо верхняя, либо нижняя граница зарплаты. Возможно, работодатели предпочитают не всегда указывать верхний и нижний пределы зарплаты при размещении вакансии, предпочитая обсуждать ее с кандидатом.
* Среднее значение для нижней границы заработной платы составляет ~71 тыс. для верхней границы - ~110 тыс.
* Большая часть вакансий предлагается для тех, кто готов работать полный рабочий день и готов к полной занятости, поскольку работодатели стремятся нанимать сотрудников, которые готовы уделять достаточно времени и энергии для выполнения поставленных задач.
* Большинство вакансий требуют опыта работы от 1 года до 3 лет. Возможно, это связано с тем, что работодатели обычно ищут кандидатов, которые уже имеют определенный уровень профессионального опыта и могут быстро адаптироваться к новой должности. 

In [752]:
# Дополнительный запрос №1 к юниту 4

# текст запроса
query_4_6 = f'''
SELECT 
    COUNT(v.id) vac_cnt,
    a.name area
FROM public.VACANCIES v
    JOIN public.AREAS a ON v.area_id = a.id
GROUP BY 2
ORDER BY 1 DESC
LIMIT 10
'''
# результат запроса
df = pd.read_sql_query(query_4_6, connection)
df


C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\3107449529.py:15: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt             area
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы
5  1698     Екатеринбург
6  1670  Нижний Новгород
7  1415           Казань
8  1301        Краснодар
9  1144           Самара

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [753]:
# текст запроса
query_5_1 = f'''
(SELECT 
    e.name employer_name,
    COUNT(v.id)
FROM public.EMPLOYERS e
    JOIN public.VACANCIES v ON v.employer_id = e.id
GROUP BY e.name
ORDER BY 2 DESC
LIMIT 1)
UNION 
(SELECT 
    e.name,
    COUNT(v.id)
FROM public.EMPLOYERS e
    JOIN public.VACANCIES v ON v.employer_id = e.id
GROUP BY e.name
order by 2 DESC
OFFSET 4 LIMIT 1)
ORDER BY 2 DESC
'''

In [754]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\3834399874.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



employer_name  count
0         Яндекс   1933
1  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [755]:
# текст запроса
query_5_2 = f'''
SELECT 
    a.name area_name,
    COUNT(DISTINCT e.id) emp_cnt,
    COUNT(DISTINCT v.id) vac_cnt
FROM public.AREAS a
    LEFT JOIN public.VACANCIES v ON v.area_id = a.id
    LEFT JOIN public.EMPLOYERS e ON e.area = a.id
WHERE v.id IS NULL
GROUP BY a.id
ORDER BY emp_cnt DESC
LIMIT 1
'''

In [756]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\3341481329.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



area_name  emp_cnt  vac_cnt
0    Россия      410        0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [757]:
# текст запроса
query_5_3 = f'''
SELECT 
    e.name employer_name,
    COUNT(DISTINCT v.area_id) vac_count
FROM public.employers e
    JOIN public.vacancies v ON v.employer_id = e.id
GROUP BY e.id
ORDER BY 2 DESC
'''

In [758]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\201667002.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



employer_name  vac_count
0                                      Яндекс        181
1                                  Ростелеком        152
2                                  Спецремонт        116
3                      Поляков Денис Иванович         88
4                                   ООО ЕФИН          71
...                                       ...        ...
14901                            НПП Авиатрон          1
14902              Центр дистанционных торгов          1
14903  Городские Телекоммуникационные Системы          1
14904                       Введенский, Отель          1
14905              Группа Компаний Автоимпорт          1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [759]:
# текст запроса
query_5_4 = f'''
SELECT 
    COUNT(e.id) emp_count
FROM public.EMPLOYERS e
    LEFT JOIN public.EMPLOYERS_INDUSTRIES emp_ind ON e.id = emp_ind.employer_id
WHERE emp_ind.industry_id IS NULL
'''

In [760]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\3493105328.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



emp_count
0       8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [761]:
# текст запроса
query_5_5 = f'''
SELECT 
    e.name employer_name
FROM public.EMPLOYERS e
    JOIN public.EMPLOYERS_INDUSTRIES emp_ind ON e.id = emp_ind.employer_id
GROUP BY 1
HAVING COUNT(emp_ind.industry_id) = 4
ORDER BY 1
OFFSET 2 LIMIT 1
'''

In [762]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\388711819.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



employer_name
0          2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [763]:
# текст запроса
query_5_6 = f'''
SELECT 
    COUNT(emp_ind.employer_id) emp_cnt
FROM public.EMPLOYERS_INDUSTRIES emp_ind
    JOIN public.INDUSTRIES i ON i.id = emp_ind.industry_id
WHERE i.name = 'Разработка программного обеспечения'
'''

In [764]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\821628714.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



emp_cnt
0     3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [765]:
# код для получения списка городов-милионников
import requests # Импортируем библиотеку requests
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup

url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России' # Определяем адрес страницы
response=requests.get(url) # Выполняем GET-запрос, содержимое ответа присваивается переменной response
page=BeautifulSoup(response.text,'html.parser') #Создаём объект BeautifulSoup, указывая html-парсер
 
cities = str(page.find('table', class_="standard")) # Извлекаем информацию из тега с классом "standard" при помощи метода find()
cities = cities.split('title="')[3:] # Разделяем строку, используя метод split()

cities_list = [] # Создаем пустой лист городов-миллионников  
for i in cities: # В цикле проходим по элементам в строке  
    cities_list.append(i.split('">')[0]) # Добавляем первый элемент строки в список 

cities = tuple(cities_list) # Создаем кортеж из списка

In [766]:
# текст запроса
query_5_7 = f'''
WITH temp AS(
SELECT 
    a.name city_name,
    COUNT(v.id) vac_cnt
FROM public.VACANCIES v
    JOIN public.AREAS a ON v.area_id = a.id
    JOIN public.EMPLOYERS e ON v.employer_id = e.id 
WHERE a.name in {cities} 
    AND e.name = 'Яндекс'
GROUP BY 1
ORDER BY 2)
SELECT *  FROM temp
UNION ALL
SELECT 
    'Total',
    sum(vac_cnt)
FROM temp
'''

In [767]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\1965596796.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



city_name  vac_cnt
0              Омск     21.0
1         Челябинск     22.0
2        Красноярск     23.0
3         Волгоград     24.0
4             Пермь     25.0
5            Казань     25.0
6    Ростов-на-Дону     25.0
7               Уфа     26.0
8            Самара     26.0
9         Краснодар     30.0
10          Воронеж     32.0
11      Новосибирск     35.0
12  Нижний Новгород     36.0
13     Екатеринбург     39.0
14  Санкт-Петербург     42.0
15           Москва     54.0
16            Total    485.0

***

# Выводы по анализу работодателей
* Яндекс предлагает наибольшее количество вакансий на рынке труда. Больше всего вакансий компания разместила в Москве, что неудивительно, так как Москва является крупным центром бизнеса и технологий в России.
* Среди регионов, где нет вакансий, Россия имеет наибольшее количество работодателей, это можно объяснить тем, что сервис ориентирован в первую очередь на российский рынок труда.
* В тройку лидеров по количеству регионов, в которых представлены вакансии, вошли такие крупные компании, как "Яндекс", "Ростелеком" и "Спецремонт".
* 8419 работодателей не указали сферу деятельности, скорее всего, работодатели предпочитают давать более подробную информацию о работе в описании вакансии, что, конечно, может затруднить поиск вакансий по ключевым словам. 
* 3553 работодателя занимаются разработкой программного обеспечения.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [768]:
# текст запроса
query_6_1 = f'''
SELECT 
    COUNT(*)
FROM public.VACANCIES
WHERE name ILIKE '%data%'
    OR name ILIKE '%данн%'
'''

In [769]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\366702258.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [770]:
# текст запроса
query_6_2 = f'''
SELECT 
    COUNT(id)
FROM public.VACANCIES
WHERE (name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name ILIKE '%ML%' and name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%')
    AND 
    (name ILIKE '%junior%'
    OR experience ILIKE '%Нет опыта%'
    OR employment ILIKE '%Стажировка%')
'''

In [771]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\2267038862.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [772]:
# текст запроса
query_6_3 = f'''
SELECT 
    COUNT(id)
FROM public.VACANCIES
WHERE (name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name LIKE '%ML%' and name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%')
    AND (key_skills ILIKE '%SQL%' OR key_skills ILIKE '%postgres%')    
'''

In [773]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\3794299702.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [774]:
# текст запроса
query_6_4 = f'''
SELECT 
    COUNT(id)
FROM public.VACANCIES
WHERE (name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name LIKE '%ML%' and name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%')
    AND (key_skills ILIKE '%Python%')    
'''

In [775]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\4264665179.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [776]:
# текст запроса
query_6_5 = f'''
SELECT 
    ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), ''))+1), 2) avg_cnt
FROM public.VACANCIES
WHERE (name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name LIKE '%ML%' and name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%')
    AND key_skills IS NOT NULL
'''

In [777]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\3654210517.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



avg_cnt
0     6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [778]:
# текст запроса
query_6_6 = f'''
SELECT 
    DISTINCT experience,
    ROUND(AVG(COALESCE((salary_from+salary_to)/2, salary_from, salary_to)))
FROM public.VACANCIES
WHERE (name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name LIKE '%ML%' and name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%')
    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY experience
'''

In [779]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\519194770.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience     round
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

***

# Выводы по предметному анализу
* Из ~49000 вакансий 1771 вакансия связана с данными. Из них 51 вакансия подходит для начинающего дата-сайентиста.
* Не все вакансии, связанные с данными, требуют знания SQL или Python. Так, например, из 1771 вакансии, имеющих отношение к  данным,  201 вакансий требуют знания SQL или Postgres, а 351 вакансий - знания Python.
* В среднем работодатели указывают 6 ключевых навыков для работы в области науки о данных.
* Чем больше опыта у дата-сайентиста, тем выше его зарплата. Возможно, это связано с тем, что с увеличением опыта дата-сайентист может стать более ценным для работодателей, так как с опытом специалист приобретает более глубокие знания и навыки, становится более профессиональным и способным решать более сложные задачи в области анализа данных.


# Общий вывод по проекту

* Больше всего вакансий в таких российских городах, как Москва, Санкт-Петербург и Новосибирск. Из этого можно сделать вывод, что чем больше население города, тем больше количество предлагаемых вакансий. Отметим, что сервис востребован и работодателями из других стран, так как в топ-5 городов по количеству вакансий вошли города из других стран.

* При размещении вакансии работодатели не всегда предпочитают указывать верхнюю и нижнюю вилки зарплаты, что может быть вызвано различными причинами, например, желанием работодателей сначала провести переговоры о зарплате с потенциальными кандидатами. Средняя зарплата указывается в диапазоне от 71 тыс. до 110 тыс. Большинство вакансий предлагают работу на полный рабочий день и на полную занятость, поскольку работодатели стремятся нанимать сотрудников, готовых уделять достаточно времени и сил для выполнения поставленных задач.

* Большинство вакансий требуют опыта работы от 1 года до 3 лет. Вероятно, это связано с тем, что работодатели обычно ищут кандидатов, которые уже имеют определенный уровень профессионального опыта и могут быстро адаптироваться к новой должности. 

* "Яндекс", "Ростелеком" и "Спецремонт" являются лидерами по количеству регионов, в которых представлены их вакансии. "Яндекс" также предлагает наибольшее количество вакансий на рынке труда. Большинство вакансий этой компании размещены в Москве, что неудивительно, так как Москва является крупным центром бизнеса и технологий в России.

* Благодаря дополнительным запросам (см. Доп.запрос №2 и №5), мы определили, что в топ-3 входят следующие вакансии: Системный администратор, Программист 1С, Бизнес-аналитик, а также то, что вакансии в сфере IT составляют большую часть из топ-10 по количеству ключевых навыков.

* Из ~49000 вакансий 1771 вакансия связана с данными. Из них 51 вакансий подходит для начинающего дата-сайентиста. Также интересно, что не все вакансии, связанные с данными, требуют знания SQL или Python. Так, например, из 1771 вакансии, имеющих отношение к  данным,  201 вакансий требуют знания SQL или Postgres, а 351 вакансий - знания Python. 

* В среднем работодатели указывают 6 ключевых навыков для работы в области науки о данных. 

* В вакансиях дата-сайентистов уровень заработной платы зависит от опыта соискателя. Это, скорее всего, связано с тем, что с увеличением опыта дата-сайентист может стать более ценным для работодателей, так как с опытом специалист приобретает более глубокие знания и навыки, становится более профессиональным и способным решать более сложные задачи в области анализа данных.

* Из дополнительного запроса №3 и графика к нему мы выявили, что большинство (45.2%) вакансий, связанных с данными, находятся в Москве. Большинство крупных IT-компаний, имеют офисы или центры разработки в Москве. Это создает большой спрос на специалистов по Data Science именно в данном городе.

* Дополнительный запрос (№4) показал, что "Сбер" является лидером по количеству вакансий, связанных с данными. Это может быть связано с тем, что "Сбер" активно развивает проекты в области искусственного интеллекта, науки о данных.

In [780]:
# Дополнительный запрос №2 
# Топ-3 вакансий:
# Текст запроса 
query_7_1 = f'''
SELECT 
    name,
    COUNT(DISTINCT id) cnt
FROM public.VACANCIES
GROUP BY 1
ORDER BY COUNT(id) DESC 
LIMIT 3
'''
# Результат запроса 
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\3908702927.py:14: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name   cnt
0  Системный администратор  1761
1           Программист 1С  1153
2          Бизнес-аналитик   488

In [785]:
# Дополнительный запрос №3 
 # Распределение data-вакансий по городам
 # Текст запроса 
query_7_2 = f'''
SELECT 
    a.name city,
    COUNT(v.id) vac_cnt
FROM public.AREAS a
    LEFT JOIN public.VACANCIES v on a.id = v.area_id
WHERE (v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR (v.name LIKE '%ML%' and v.name NOT ILIKE '%HTML%')
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%')
GROUP BY 1
ORDER BY 2 DESC
'''
 # Результат запроса
df = pd.read_sql_query(query_7_2, connection)
df

# Создаем функцию для определения города 
def get_city(arg):
    # Список с городами-миллионниками
    million_cities = ['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород','Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']
    # Разделяем строку на отдельные элементы при помощи метода split() и разделителя, выделяем первый элемент - город
    arg_splitted = arg.split(' , ')[0]
    # Проверяем условие, если город находится в списке
    if arg_splitted in million_cities:
        # Возвращаем город
        return arg_splitted
    else:
        # Если условие не выполняется, возвращаем 'Другие'
        return 'Другие'
    
# Применяем функцию к столбцу 'city'  
df['city'] = df['city'].apply(get_city) 
display(df)

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\139481534.py:20: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



city  vac_cnt
0            Москва      217
1   Санкт-Петербург       64
2       Новосибирск       23
3   Нижний Новгород       20
4            Казань       16
5            Другие       16
6            Другие        9
7            Другие        8
8      Екатеринбург        7
9         Краснодар        6
10   Ростов-на-Дону        6
11           Другие        6
12           Другие        6
13           Самара        5
14            Пермь        5
15           Другие        5
16           Другие        4
17           Другие        4
18           Другие        3
19           Другие        3
20           Другие        3
21           Другие        3
22             Омск        3
23              Уфа        2
24           Другие        2
25           Другие        2
26           Другие        2
27           Другие        2
28           Другие        2
29           Другие        2
30           Другие        2
31          Воронеж        2
32           Другие        1
33           Другие        1
34           Другие        1
35           Другие        1
36           Другие        1
37           Другие        1
38        Волгоград        1
39           Другие        1
40           Другие        1
41           Другие        1
42       Красноярск        1
43           Другие        1
44           Другие        1
45           Другие        1
46           Другие        1
47           Другие        1
48           Другие        1
49           Другие        1
50        Челябинск        1
51           Другие        1

In [782]:
# График к дополнительному запросу №3 
import plotly.express as px
import matplotlib.pyplot as plt

fig=plt.figure(figsize=(10,7))
fig = px.pie(
    data_frame=df,
    names = 'city',
    values = 'vac_cnt',
    title='DS vacancies distribution by city'
)
fig.show()


<Figure size 1000x700 with 0 Axes>

<image src="plotly/img1.png" alt="DS vacancies distribution by city">

In [783]:
# Дополнительный запрос №4 
# Топ 5 работодателей по количеству data-вакансий
 # Текст запроса 
query_7_3 = f'''
SELECT 
    e.name emp_name,
    COUNT(v.id) vac_cnt
FROM public.EMPLOYERS e 
    JOIN public.VACANCIES v ON e.id = v.employer_id
WHERE (v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR (v.name LIKE '%ML%' and v.name NOT ILIKE '%HTML%')
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%')
GROUP BY 1
ORDER BY 2 DESC
LIMIT 5
'''
# Результат запроса
df = pd.read_sql_query(query_7_3, connection)
df


C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\464826809.py:21: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



emp_name  vac_cnt
0                   СБЕР       37
1        Bell Integrator       25
2         Банк ВТБ (ПАО)       18
3                     VK       15
4  Positive Technologies       11

In [784]:
# Дополнительный запрос №5 
# # Топ 10 вакансий по количеству ключевых навыков
 # Текст запроса 
query_7_4 = f'''
SELECT 
    name vac_name, 
    (LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), ''))+1) key_skills_cnt,
    COUNT(id)
FROM public.VACANCIES
WHERE key_skills is NOT NULL
GROUP BY 1, 2
ORDER BY 2 DESC, 3 DESC
LIMIT 15
'''
# Результат запроса
df = pd.read_sql_query(query_7_4, connection)
df

C:\Users\mariy\AppData\Local\Temp\ipykernel_14396\1299652936.py:16: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



vac_name  key_skills_cnt  count
0            Mechanical Engineer (Rotating Equipment)              30      3
1                             Системный администратор              30      3
2                          ИТ-архитектор (enterprise)              30      2
3   Продуктовый аналитик (Product Analyst/ Data An...              30      1
4                                    Менеджер проекта              30      1
5   Специалист по информационной безопасности (pen...              30      1
6                              Программист (удаленно)              30      1
7                    Преподаватель IT курсов (онлайн)              30      1
8                          Менеджер интернет-проектов              30      1
9                               Обработчик чатов OZON              30      1
10                            Программист 1С - стажер              30      1
11                               Координатор проектов              30      1
12                                      IT-специалист              30      1
13                 Помощник системного администратора              30      1
14                    Ведущий системный администратор              30      1